fix: Fix ApkInstaller for Android <7 (SDK version <24)

gmaclennan opened the pull request:
Fixes #582 

It turns out that for Android <7, the APK and any parent folders need to have their file permissions set be world-readable. E.g. we store the APK in `[INTERNAL_FILES_DIR]/installer/myApk.apk`. To be able to install this, `myApk.apk` needs to be world-readable and the `installer` folder needs to be world-executable.

We _could_ set these permissions when creating the `installer` folder, but since these permissions are only needed on Android <7 I think it's better to leave the folder with more restrictive permissions, and set the permissions in the Java code which only runs for Android <7.

Ideally the Java code that sets permissions would set all parent folders up to `[INTERNAL_FILES_DIR]` to be world-executable, but my Java knowledge it not good enough, so for now it assumes the APK is in a folder one-level deep, e.g. it only sets permissions on the parent folder.

@luandro can you test this on your phone and see if it now works as expected?

gmaclennan committed:
fix: Fix ApkInstaller for Android <7 (SDK version <24)

gmaclennan commented:
FYI the changes to the Java code here are mainly copied from F-Droid from:

- https://gitlab.com/cde/fdroidclient/-/blob/master/app/src/main/java/org/fdroid/fdroid/installer/ApkFileProvider.java
- https://gitlab.com/cde/fdroidclient/-/blob/master/app/src/main/java/org/fdroid/fdroid/installer/DefaultInstallerActivity.java

I also borrowed some code from https://github.com/mikehardy/react-native-update-apk

luandro commented:
Great job @gmaclennan, it's working now!

I was going to point out the code for installing apk from [react-native-send-intent](https://github.com/lucasferreira/react-native-send-intent/blob/master/android/src/main/java/com/burnweb/rnsendintent/RNSendIntentModule.java#L416), but you handled it.

I suppose we're just missing:
- [ ] cleaning up logs
- [ ] not changing permissions when Android > 6

gmaclennan commented:
rebased on `share-apk-p2p`.

> I suppose we're just missing:
> 
> * [ ]  cleaning up logs

@luandro not sure what you mean. I added logging intentionally for any errors. I think it should be left in.

> * [ ]  not changing permissions when Android > 6

This is already here - the permission change only happens if `Build.VERSION.SDK_INT < Build.VERSION_CODES.N`

I think this is good to merge, do you agree @luandro?

luandro commented:
The logging makes sense @gmaclennan. I was reading the `if` statement wrong, you're right. Ok all good to merge.

hackergrrl commented:
🚀
